game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -61,71 +97,97 @@ public void retreatUnits(final IDelegateBridge bridge) {
 
-    if (battleState.isAmphibious()) {
-      retreatData = getAmphibiousRetreatData();
-      if (retreatData == null) {
-        return;
-      }
+    Retreater retreater = null;
 
+    if (battleState.isAmphibious()) {
+      retreater = getAmphibiousRetreater();
     } else if (canAttackerRetreat()) {
-      retreatData =
-          RetreatData.of(RetreatType.DEFAULT, battleState.getAttackerRetreatTerritories());
-    } else {
-      return;
+      retreater = new RetreaterGeneral(battleState);
     }
 
-    battleActions.queryRetreat(false, retreatData.retreatType, bridge, retreatData.retreatSites);
+    if (retreater != null) {
+      retreat(bridge, retreater);
+    }
   }
 
-  private @Nullable RetreatData getAmphibiousRetreatData() {
+  private Retreater getAmphibiousRetreater() {
     if (canAttackerRetreatPartialAmphib()) {
-      return RetreatData.of(
-          RetreatType.PARTIAL_AMPHIB, battleState.getAttackerRetreatTerritories());
-
+      return new RetreaterPartialAmphibious(battleState);
     } else if (canAttackerRetreatAmphibPlanes()) {
-      return RetreatData.of(RetreatType.PLANES, Set.of(battleState.getBattleSite()));
-
-    } else {
-      return null;
+      return new RetreaterAirAmphibious(battleState);
     }
+    return null;
   }
 
-  @Value(staticConstructor = "of")
-  private static class RetreatData {
-    RetreatType retreatType;
-    Collection<Territory> retreatSites;
-  }
+  private void retreat(final IDelegateBridge bridge, final Retreater retreater) {
+    final Collection<Unit> retreatUnits = retreater.getRetreatUnits();
+    final Collection<Territory> possibleRetreatSites =
+        retreater.getPossibleRetreatSites(retreatUnits);
+    final String text = retreater.getQueryText();
+
+    bridge.getDisplayChannelBroadcaster().gotoBattleStep(battleState.getBattleId(), getName());
+    final Territory retreatTo =
+        battleActions.queryRetreatTerritory(
+            battleState, bridge, battleState.getAttacker(), possibleRetreatSites, text);
+    if (retreatTo == null) {
+      return;
+    }
+    if (!battleState.isHeadless()) {
+      SoundUtils.playRetreatType(
+          battleState.getAttacker(), retreatUnits, retreater.getRetreatType(), bridge);
+    }
 
-  private boolean canAttackerRetreat() {
-    return RetreatChecks.canAttackerRetreat(
-        battleState.getUnits(BattleState.Side.DEFENSE),
-        battleState.getGameData(),
-        battleState::getAttackerRetreatTerritories,
-        battleState.isAmphibious());
-  }
+    final CompositeChange change = new CompositeChange();
 
-  private boolean canAttackerRetreatSeaPlanes() {
-    return battleState.getBattleSite().isWater()
-        && battleState.getUnits(BattleState.Side.OFFENSE).stream().anyMatch(Matches.unitIsAir());
+    change.add(retreater.extraRetreatChange(retreatTo, retreatUnits));
+
+    final Map<Retreater.RetreatLocation, Collection<Unit>> retreatingUnitMap =
+        retreater.splitRetreatUnits(retreatUnits);
```

Comment:
I've thought about having the retreaters have a `getChanges()` method that returns a list of change objects.  These change objects would encapsulate the `retreatSameTerritoryUnits`, `retreatOtherTerritoryUnits`, and `extraRetreatChange` logic.  But I didn't think the existing `Change` objects fit this.  The `HistoryNode` objects are kind of a good fit since these "changes" already create history nodes but `HistoryNode` doesn't seem to have actions to perform.

The api of this "change" object would be something like:
```
interface BattleChange {
  void writeHistoryNode();
  Change perform();
}
```
And `perform` would build the actual `Change` objects as well as modify the `BattleState` (if needed).